ledger-core/v2/scripts/gen_plantuml.go

```diff
@@ -0,0 +1,562 @@
+                uml += fmt.Sprintf("\n%s -[#DarkGreen]-> %s : RepeatOrJump(Else)", state.Name, ret.Args[3].Fun)
+            default:
+                pf.diag("\n(=> (. %s %s)", ret.Var.Obj, ret.Var.Fun)
+                for _, arg := range ret.Args {
+                    pf.diag("\n       %s", fmt.Sprintf("(. %s %s)", arg.Obj, arg.Fun))
+                }
+                pf.diag(")")
+            }
+        }
+    }
+    uml += "\n@enduml\n"
+    return uml
+}
+
+func ParseFile(fileName string, dbg ...bool) *ParsedFile {
+    pf := &ParsedFile{
+        filename: fileName,
+        dbg:      dbg[0],
+    }
+
+    sourceCode, err := slurpFile(fileName)
+    if err != nil {
+        return nil
+    }
+    pf.code = sourceCode
+
+    pf.fileSet = token.NewFileSet()
+    node, err := parser.ParseFile(pf.fileSet, pf.filename, pf.code, parser.ParseComments)
+    if err != nil {
+        return nil
+    }
+    pf.node = node
+
+    pf.states = make(map[string]*FnState)
+
+    ast.Inspect(node, func(n ast.Node) bool {
+        fn, ok := n.(*ast.FuncDecl)
+        if ok {
+            pf.parseMethod(fn)
+        }
+        return true
+    })
+
+    return pf
+}
+
+func writeUml(path string, uml string)  {
+    name := filepath.Base(path)
+    name = strings.Replace(name, ".go", "", -1)+".plantuml"
+    umlPath := fmt.Sprintf("%s/%s/%s", os.Getenv("GOPATH"), filepath.Dir(path), name)
+
+    file, err := os.Create(umlPath)
+    if err != nil {
+        fmt.Printf("Failed to create file: %s\n", umlPath)
+        return
+    }
+
+    defer file.Close()
+
+    _, err = file.WriteString(uml)
+    if err != nil {
+        fmt.Printf("Failed to write file: %s\n", umlPath)
+        return
+    }
+
+    err = file.Sync()
+    if err != nil {
+        fmt.Printf("Failed to sync file: %s\n", umlPath)
+        return
+    }
+
+    fmt.Printf("Uml saved: %s", umlPath)
+}
+
+func slurpFile(fileName string) ([]byte, error) {
+    file, err := os.OpenFile(fileName, os.O_RDONLY, 0)
+    if err != nil {
+        return nil, errors.New(fmt.Sprintf("Can't open file: [%s]", filename))
+    }
+    defer file.Close() //nolint: errcheck
+
+    res, err := ioutil.ReadAll(file)
+    if err != nil {
+        return nil, errors.New(fmt.Sprintf("Can't read file: [%s]", filename))
+    }
+    return res, nil
+}
+
+func (pf *ParsedFile) parseMethod(fn *ast.FuncDecl) {
+
+    // I want to analise only method functions (if exists)
+    if nil == fn.Recv {
+        pf.diag("\n:parseMethod: skip %s - No receiver", fn.Name.Name)
+    } else {
+
+        for _, fld := range fn.Recv.List {
+
+            // Receiver
+            recv := &RecvPair{
+                Name: fld.Names[0].Name,
+                Type: fmt.Sprintf("%s", pf.code[fld.Type.Pos()-1:fld.Type.End()-1]),
+            }
+
+            // Parameters
+            pars := make(map[string]string, 0)
+            for _, par := range fn.Type.Params.List {
+                if nil == par.Names {
```

Comment:
this could be written as 
```
name := "unnamed-param"
if par.Names != nil {
    name = par.Names[0].Name
}
pars[name] = fmt.Sprintf("%s", pf.code[par.Type.Pos()-1:par.Type.End()-1])
```
which is easier to read